src/disco/containers/InstallButton.js

```diff
@@ -29,12 +32,55 @@ export class InstallButton extends React.Component {
     downloadProgress: 0,
   }
 
+  componentDidMount() {
+    const { dispatch, guid, installURL, slug } = this.props;
+    this.addonManager = new AddonManager(guid, installURL, this.statusChanged);
+    this.addonManager
+      .getAddon()
+      .then(() => {
+        dispatch({
+          type: 'INSTALL_STATE',
+          payload: {slug, guid, url: installURL, status: INSTALLED}});
+      }, () => {
+        dispatch({
+          type: 'INSTALL_STATE',
+          payload: {slug, guid, url: installURL, status: UNINSTALLED}});
+      });
+  }
+
+  statusChanged = (addonInstall) => {
+    const { dispatch, slug } = this.props;
+    if (addonInstall.state === 'STATE_DOWNLOADING') {
+      const downloadProgress = parseInt(
+        100 * addonInstall.progress / addonInstall.maxProgress, 10);
+      dispatch({type: 'DOWNLOAD_PROGRESS', payload: {slug, downloadProgress}});
+    } else if (addonInstall.state === 'STATE_INSTALLING') {
+      dispatch({type: 'START_INSTALL', payload: {slug}});
+    } else if (addonInstall.state === 'STATE_INSTALLED') {
+      dispatch({type: 'INSTALL_COMPLETE', payload: {slug}});
+    }
+  }
+
+  install() {
+    const { dispatch, slug } = this.props;
+    dispatch({type: 'START_DOWNLOAD', payload: {slug}});
+    this.addonManager.install();
+  }
+
+  uninstall() {
+    const { dispatch, slug } = this.props;
+    dispatch({type: 'START_UNINSTALL', payload: {slug}});
+    this.addonManager
```

Comment:
This is essentially private since it's a react component so I don't think we need to. The component itself doesn't use the promises since it just waits for react to re-render it with data changes.